api/cpp/include/slint.h

```diff
@@ -1049,6 +1049,179 @@ class MapModel : public Model<MappedModelData>
     std::function<MappedModelData(const SourceModelData &)> map_fn;
 };
 
+template<typename ModelData>
+class SortModel;
+
+namespace private_api {
+template<typename ModelData>
+struct SortModelInner : private_api::ModelChangeListener
+{
+    SortModelInner(std::shared_ptr<slint::Model<ModelData>> source_model,
+                   std::function<bool(const ModelData &, const ModelData &)> sort_fn,
+                   slint::SortModel<ModelData> &target_model)
+        : source_model(source_model), sort_fn(sort_fn), target_model(target_model)
+    {
+    }
+
+    void row_added(int first_inserted_row, int count) override
+    {
+        if (sorted_rows_dirty) {
+            reset();
+            return;
+        }
+
+        // Adjust the existing sorted row indices to match the updated source model
+        for (auto it = sorted_rows.begin(), end = sorted_rows.end(); it != end; ++it) {
+            if (*it >= first_inserted_row)
+                *it += count;
+        }
+
+        for (int row = first_inserted_row; row < first_inserted_row + count; ++row) {
+
+            ModelData inserted_value = *source_model->row_data(row);
+            auto insertion_point =
+                    std::lower_bound(sorted_rows.begin(), sorted_rows.end(), inserted_value,
+                                     [this](int sorted_row, const ModelData &inserted_value) {
+                                         auto sorted_elem = source_model->row_data(sorted_row);
+                                         return sort_fn(*sorted_elem, inserted_value);
+                                     });
+
+            insertion_point = sorted_rows.insert(insertion_point, row);
+            target_model.row_added(insertion_point - sorted_rows.begin(), 1);
```

Comment:
Isen't this what `std::distance` is meant to be used for?

Reply:
That's a good idea. I guess for vector iterators it's not strictly needed, but I agree that it makes sense to use it here (and in the other spots), if just to increase readability of code.